proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,652 @@
+.. proposal-number:: Leave blank. This will be filled in when the proposal is
+                     accepted.
+
+.. trac-ticket:: Leave blank. This will eventually be filled with the Trac
+                 ticket number which will track the progress of the
+                 implementation of the feature.
+
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+
+.. highlight:: haskell
+
+This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/0>`_. **After creating the pull request, edit this file again, update the number in the link, and delete this bold sentence.**
+
+.. contents::
+
+Linear Types
+============
+
+This proposal introduces a notion of *linear function* to GHC. Linear
+functions are regular functions that guarantee that they will use
+their argument exactly once. Whether a function ``f`` is linear or not
+is called the *multiplicity* of ``f``. We propose a new language
+extension, ``-XLinearTypes``. When turned on, the user can enforce
+a given multiplicity for ``f`` using a type annotation.
+
+The theory behind this proposal has been fully developed in a peer
+reviewed conference publication that will be presented at POPL'18. See
+the `extended version of the paper
+<https://arxiv.org/abs/1710.09756>`_.
+
+Motivation
+----------
+
+Haskell, along with a few other languages, heralded the notion of
+*type safety* into mainstream programming. That is, *well-typed
+programs do not go wrong*. Well-typed programs do sometimes crash, or
+fail to terminate, but they do not segfault. Now, the system resources
+that these programs manipulate have changing states, need to be
+initialized before use and conversely, must be freed in a timely
+manner. We want not just type safety in Haskell, but also *resource
+safety*. We want well-typed programs that do not go wrong in the sense
+that they might still crash, but they do not rewind the state of I/O
+resources, these resources are never used before they are initialized,
+are guaranteed to be freed by the time control flow exits user defined
+scopes, and never used after being freed.
+
+This proposal hits another goal as a side benefit. In Haskell, impure
+computations are typically structured as a sequence of steps, be it in
+the ``IO`` monad or in ``ST``. The latter in particular serves to
+precisely control which effects are possible and the scope within
+which they are visible. But using monads to write "locally impure"
+computations that still look pure from the outside has an unfortunate
+consequence: computations are oversequentialized, making it hard for
+the compiler to recover lost opportunities for parallelism.
+
+Linear types enable better solutions to both problems:
+
+1. using types to guarantee resource safety, and
+2. using types to control the scope of effects without forcing an
+   unnatural sequencing of mutually independent effects.
```

Comment:
The benefits of linear types are sprinkled through all this section. It could make sense to collect them all in one list, and then discuss them with examples. From what I can grok:
```
Linear types enable better solutions to:
1. using types to guarantee resource safety,
2. increasing the opportunities for parallelism in effectful computations,
3. reducing latency due to GC calls
```
The Array example covers (1) and (2), the malloc/free and the socket examples cover (1). Aren't we missing here an example of (3)?

Reply:
We already have an example for (3): it's the `malloc/free` stuff. I don't think a more elaborate example would offer any new insight that a good description wouldn't provide. Note also that (3) is not quite true. Linear types don't give you (3). You can get (3) without them (using manual memory management). It's just that (1) then becomes much more important (because memory becomes a system resource you have to manage like any other).

Reply:
Well, it is claiming that it helps, not really that it makes it possible. How does it help? For instance by not requiring to peek and poke Haskell values to keep them out of the gc heap.

Reply:
I don't see such a claim anywhere. Quote:

> Systems programming with quasi real-time requirements can often benefit from easing pressure on the GC by taking long-lived objects out of the GC-managed heap entirely. Fewer long-lived objects in the heap means faster major collection times, hence shorter GC pauses. Linear types enable *safe* manual memory management for long-lived objects.

We *do* need to peek and poke Haskell values if kept out of the GC heap. All we're doing here is squarely item (1) that you list. No more no less. We don't help in reducing latencies. We help in making resources safe to manage. If memory becomes an explicitly managed resource, as one sometimes does when trying to reduce max latencies, then resource safety is certainly important. That's still (1).

Reply:
I must be outdated in my readings. Former versions of the paper discussed a linear heap, which is not garbage collected.